File: Solutions/C++/0006-ZigzagConversion/ZigZagConversion.cpp
#include <string>
using namespace std;

class Solution {
public:
    string convert(string s, int numRows)
    {
        if(numRows == 1 || s.length() == 1)
            return s;

        string res = "";
        for(int i = 0; i < numRows; i++) {

            int pos = i;
            bool flip = (i == numRows - 1 ? true : false);
            while(pos < s.length()) {
                res += s[pos];
                if(flip)
                    pos += (i + 1) * 2 - 2;
                else
                    pos += (numRows - i) * 2 - 2;

                if(i != 0 && i != numRows - 1)
                    flip = !flip;
            }
        }

        return res;
    }
};

File: Solutions/C#/0006-ZigZagConversion/ZigZagConversion.cs
using System;

namespace ZigZagConversion
{
    public class Solution
    {
        public string Convert(string s, int numRows)
        {
            if (numRows == 1 || s.Length < numRows)
                return s;

            string ans = "";
            bool flag;

            for (int i = 0; i < numRows; i++)
            {
                flag = false;
                int j = i;
                while (j < s.Length)
                {
                    ans += s[j];
                    if (i == 0 || i == numRows - 1)
                        j += 2 * numRows - 2;

                    else
                    {
                        if (!flag)
                            j += 2 * numRows - 2 * i - 2;
                        else
                            j += 2 * i;

                        flag = !flag;
                    }
                }
            }

            return ans;
        }
    }
}
